File: clases/Cap03_Seleccion_de_Caracteristicas/balu/main.py
from my_utils import dirfiles, num2fixstr, extract_features, Accuracy
import sys
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from pybalu.feature_selection import clean # > pip3 install scipy==1.2
from pybalu.feature_selection import sfs
from pybalu.feature_transformation import normalize

# Extracting features
X0_train = extract_features('Training_0','png')
X1_train = extract_features('Training_1','png')
X0_test  = extract_features('Testing_0','png')
X1_test  = extract_features('Testing_1','png')


# Training dataset
print('Training Subset:')
X_train  = np.concatenate((X0_train,X1_train),axis=0)
d0_train = np.zeros([X0_train.shape[0],1],dtype=int)
d1_train = np.ones([X1_train.shape[0],1],dtype=int)
d_train  = np.concatenate((d0_train,d1_train),axis=0)
print('Original extracted features: '+str(X_train.shape[1])+ '('+str(X_train.shape[0])+' samples)')

# Training: Cleaning
sclean = clean(X_train,show=True)
X_train_clean = X_train[:,sclean]
print('           cleaned features: '+str(X_train_clean.shape[1])+ '('+str(X_train_clean.shape[0])+' samples)')

# Training: Normalization
X_train_norm, a, b = normalize(X_train_clean)
print('        normalized features: '+str(X_train_norm.shape[1])+ '('+str(X_train_norm.shape[0])+' samples)')

# Training: Feature selection
ssfs = sfs(X_train_norm, d_train, n_features=20 ,method="fisher", show=True)
X_train_sfs = X_train_norm[:,ssfs]
print('          selected features: '+str(X_train_sfs.shape[1])+ '('+str(X_train_sfs.shape[0])+' samples)')

# Testing dataset
print('Testing Subset:')
X_test  = np.concatenate((X0_test,X1_test),axis=0)
d0_test = np.zeros([X0_test.shape[0],1],dtype=int)
d1_test = np.ones([X1_test.shape[0],1],dtype=int)
d_test  = np.concatenate((d0_test,d1_test),axis=0)

# Testing: Cleaning
X_test_clean = X_test[:,sclean]

# Testing: Normalization
X_test_norm = X_test_clean * a + b

# Testing: Feature selection
X_test_sfs = X_test_norm[:,ssfs]
print('    clean+norm+sfs features: '+str(X_test_sfs.shape[1])+ '('+str(X_test_sfs.shape[0])+' samples)')

# Classification on Testing dataset
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_sfs, d_train)
ds = knn.predict(X_test_sfs)
Accuracy(ds, d_test,'Testing')

File: clases/Cap03_Seleccion_de_Caracteristicas/balu/main.m
% Extracting features
X0_train = extract_features('Training_0','png');
X1_train = extract_features('Training_1','png');
X0_test  = extract_features('Testing_0','png');
X1_test  = extract_features('Testing_1','png');

% Training dataset
disp('Training Subset:')
X_train  = [X0_train;X1_train];
d0_train = zeros(size(X0_train,1),1);
d1_train = ones(size(X1_train,1),1);
d_train  = [d0_train;d1_train];
fprintf('Original extracted features: %d (%d samples)\n',size(X_train,2),size(X_train,1));

% Training: Cleaning
sclean        = Bfs_clean(X_train);
X_train_clean = X_train(:,sclean);
fprintf('          cleaned features: %d (%d samples)\n',size(X_train_clean,2),size(X_train_clean,1));

% Training: Normalization
[X_train_norm, a, b] = Bft_norm(X_train_clean,0);
fprintf('       normalized features: %d (%d samples)\n',size(X_train_norm,2),size(X_train_norm,1));

% Training: Feature selection
opsfs.m      = 20;                  % features to be selected
opsfs.show   = 1;                   % display/no display results
opsfs.b.name = 'fisher';            % SFS with Fisher
ssfs         = Bfs_sfs(X_train_norm,d_train,opsfs);
X_train_sfs  = X_train_norm(:,ssfs);
fprintf('         selected features: %d (%d samples)\n',size(X_train_sfs,2),size(X_train_sfs,1));

% Testing dataset
disp('Testing Subset:')
X_test  = [X0_test;X1_test];
d0_test = zeros(size(X0_test,1),1);
d1_test =  ones(size(X1_test,1),1);
d_test  = [d0_test;d1_test];

% Testing: Cleaning
X_test_clean = X_test(:,sclean);

% Testing: Normalization
N = size(X_test_clean,1);
X_test_norm = X_test_clean.*(ones(N,1)*a) + ones(N,1)*b;

% Testing: Feature Selection
X_test_sfs = X_test_norm(:,ssfs);
fprintf('   clean+norm+sfs features: %d (%d samples)\n',size(X_test_sfs,2),size(X_test_sfs,1));


% Classification on Testing dataset
opknn.k = 5; % number of neighbors
ds      = Bcl_knn_old(X_train_sfs,d_train,X_test_sfs,opknn);
C       = Bev_confusion(ds, d_test);
acc     = Bev_performance(ds, d_test);
disp('Confusion Matrix:')
C
fprintf('Accuracy = %f\n',acc);





